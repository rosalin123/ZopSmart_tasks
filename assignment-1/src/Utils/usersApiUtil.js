const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchUsers = async () => {
  try {
    let url = `${baseUrl}/users`;
    let users = await fetch(url, { method: 'GET' });
    let usersJson = await users.json();
    return usersJson;
  } catch (err) {
    console.log('Error fetching users details', err.stack);
  }
};

export const getUser = async (id) => {

    let url = `${baseUrl}/users/${id}`;
    let user = await fetch(url, { method: 'GET' });
    let userJson = await user.json();
    return userJson;

};
