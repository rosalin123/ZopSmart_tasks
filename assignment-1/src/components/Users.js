import React, { Component } from 'react';
import { fetchUsers } from '../actions/userActions';
import { connect } from 'react-redux';
import { Grid } from '@material-ui/core';
import UserComponent from './UserComponent';
import { withStyles } from '@material-ui/core/styles';
import { clearUserPosts } from '../actions/postActions';
import Loader from './loader';
import ErrorComponent from './error';

const styles = {
  root: {
    paddingTop: '10px',
  },
  userStyles: {
    borderBottom: '1px solid lightgrey',
    background: 'white',
  },
};

export class Users extends Component {
  componentDidMount = () => {
    this.props.fetchUsers();
    this.props.clearUserPosts();
  };

  render() {
    const { classes } = this.props;
    const { users, loading, error } = this.props;
    return loading === true ? (
      <Loader data-test="loaderComponent" />
    ) : error !== '' ? (
      <ErrorComponent
        message="Something's not right.."
        data-test="errorComponent"
      />
    ) : (
      <Grid container spacing={3} direction="column" className={classes.root}>
        {users.map((user, index) => {
          return (
            <Grid item key={index} className={classes.userStyles}>
              <UserComponent
                username={user.username}
                email={user.email}
                id={user.id}
              />
            </Grid>
          );
        })}
      </Grid>
    );
  }
}

export const mapStateToProps = (state) => ({
  loading: state.Users.loading,
  error: state.Users.error,
  users: state.Users.users,
});

export const mapDispatchToProps = (dispatch) => {
  return {
    fetchUsers: () => dispatch(fetchUsers()),
    clearUserPosts: () => dispatch(clearUserPosts()),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withStyles(styles)(Users));
