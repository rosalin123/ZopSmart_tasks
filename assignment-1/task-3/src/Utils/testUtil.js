import React from 'react';

export const findbyTestAttr = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};
