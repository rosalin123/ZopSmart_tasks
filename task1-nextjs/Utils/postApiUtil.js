import axios from 'axios';
import fetch from 'isomorphic-unfetch';
const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchPosts = async () => {
  let url = `${baseUrl}/posts`;
  let posts = await axios.get(url);
  let postsJson = await posts.json();
  return postsJson;
};

export const getPost = async (id) => {
  let url = `${baseUrl}/posts/${id}`;
  let post = fetch(url);
  let postJson = await post.json();
  return postJson;
};

export const getPostComments = async (id) => {
  let url = `${baseUrl}/posts/${id}/comments`;
  let comments = await axios.get(url);
  let commentsJson = await comments.json();
  return commentsJson;
};

export const getUserPosts = async (id) => {
  let url = `${baseUrl}/posts?userId=${id}`;
  let userPosts = await fetch(url);
  let userPostsJson = await userPosts.json();
  return userPostsJson;
};
