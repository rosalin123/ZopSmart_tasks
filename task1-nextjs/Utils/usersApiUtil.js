import fetch from 'isomorphic-unfetch';
const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchUsers = async () => {
  try {
    let url = `${baseUrl}/users`;
    let users = await fetch(url);
    let usersJson = await users.json();
    return usersJson;
  } catch (err) {
    console.log('Error fetching users details', err.stack);
  }
};

export const getUser = async (id) => {
  let url = `${baseUrl}/users/${id}`;
  let user = await fetch(url);
  let userJson = await user.json();
  return userJson;
};
