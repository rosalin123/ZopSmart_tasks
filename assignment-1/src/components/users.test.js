import React from 'react';
import { render, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import Users, { Users as UsersComponent } from './Users';
import { mapDispatchToProps } from './Users';
import { findbyTestAttr } from '../Utils/testUtil';

const mockStore = configureMockStore();

const setUp = (initialState = {}) => {
  const store = mockStore(initialState);
  const component = render(
    <Router>
      <Provider store={store}>
        <Users />
      </Provider>
    </Router>
  );
  return component;
};

describe('Users Component', () => {
  let component;
  it('Users component should render without throwing an error', () => {
    const initialState = {
      Users: {
        loading: false,
        users: [{ username: 'ros', email: 'r@gmail.com', id: 1 }],
        error: '',
      },
    };
    component = setUp(initialState);
    expect(component).toMatchSnapshot();
  });

  it('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).fetchUsers();
    mapDispatchToProps(dispatch).clearUserPosts();
    expect(typeof dispatch.mock.calls[0][0]).toBe('function');
    expect(typeof dispatch.mock.calls[1][0]).toBe('function');
  });

  it('componentDidMount calls methods as expected', () => {
    const props = {
      loading: false,
      error: '',
      users: [],
      classes: { root: {}, userStyles: {} },
      fetchUsers: jest.fn(),
      clearUserPosts: jest.fn(),
    };

    component = shallow(<UsersComponent {...props} />);
    const instance = component.instance();

    jest.spyOn(instance.props, 'fetchUsers');
    jest.spyOn(instance.props, 'clearUserPosts');
    instance.componentDidMount();
    expect(instance.props.fetchUsers).toHaveBeenCalledTimes(1);
    expect(instance.props.clearUserPosts).toHaveBeenCalledTimes(1);
  });

  it('loader should render when loading is true', () => {
    const props = {
      loading: true,
      error: '',
      users: [],
      classes: { root: {}, userStyles: {} },
      fetchUsers: jest.fn(),
      clearUserPosts: jest.fn(),
    };

    component = shallow(<UsersComponent {...props} />);
    let wrapper = findbyTestAttr(component, 'loaderComponent');
    expect(wrapper.length).toBe(1);
  });

  it('ErrorComponent should be rendered when there is any error', () => {
    const props = {
      loading: false,
      error: 'Something wrong',
      users: [],
      classes: { root: {}, userStyles: {} },
      fetchUsers: jest.fn(),
      clearUserPosts: jest.fn(),
    };

    component = shallow(<UsersComponent {...props} />);
    expect(findbyTestAttr(component, 'errorComponent').length).toBe(1);
    expect(findbyTestAttr(component, 'loaderComponent').length).toBe(0);
  });
});
