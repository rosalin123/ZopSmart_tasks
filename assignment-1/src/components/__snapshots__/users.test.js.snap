// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Component Users component should render without throwing an error 1`] = `
<div
  class="MuiGrid-root Users-root-1 MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-direction-xs-column"
>
  <div
    class="MuiGrid-root Users-userStyles-2 MuiGrid-item"
  >
    <div
      class="MuiGrid-root makeStyles-root-3 MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-align-items-xs-center"
      data-test="userComponent"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
        data-test="avatar"
      >
        <a
          class="makeStyles-linkStyles-6"
          href="/users/1"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-4 MuiAvatar-colorDefault"
          >
            r
          </div>
        </a>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
        data-test="userInfo"
      >
        <a
          class="makeStyles-linkStyles-6"
          href="/users/1"
        >
           
          <p
            class="MuiTypography-root makeStyles-typographyStyles-5 MuiTypography-body1"
          >
            ros
          </p>
           
        </a>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          r@gmail.com
        </p>
      </div>
    </div>
  </div>
</div>
`;
